`Theme`: Use intersection observer to render a placeholder when not visible

xavier-lc opened the pull request:
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Related to https://github.com/Automattic/wp-calypso/issues/95248

## Proposed Changes

* Update the `theme` component so it's lazy loaded when it's visible or within the next viewport height.

## Why are these changes being made?
<!--
It's easy to see what a PR does but much harder to find out why it was made,
particularly when researching old changes in history. Record an explanation of
the motivation behind this change and how it will help.
-->

* The `/themes` page loads ~34MB of theme screenshots on page load, because it's rendering 100 themes for each page. This is too much, and lazy loading the component allows us to reduce that to ~5MB.

* The initial [PR](https://github.com/Automattic/wp-calypso/pull/95881) had to be [reverted](https://github.com/Automattic/wp-calypso/pull/95951) because it was breaking an E2E test, which I fixed on https://github.com/Automattic/wp-calypso/pull/95984/commits/562d8b9afb871a3c535e0a469dc53850e8942c4a.

* I also noticed that the lazy loading was not working properly on the logged in theme showcase, with only the visible themes being loaded. This happened because the intersection observer's default root element is the document viewport, and since the logged in theme showcase has its own scroll, the intersection is not properly calculated. In order to fix it, I provided the theme showcase wrapper ref through a react context.

## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* On the `Files changed` tab, enable the `Hide whitespace` option so the diff is more readable

* Open the live preview while logged out
* Open the dev tools, go to the network tab, filter by `.png`
* Go to `/themes`
* Check that only 8 theme screenshots are loaded (could be more, depending on the screen size), resulting in way less image data transferred (`5.1MB`) than what's happening on prod (`33.6MB`):
* Scroll down, check that the theme images are loaded 
* Check that it works as well for mobile devices
* Repeat the previous steps on the logged in theme showcase

|Before|After|
|---|---|
|![Untitled](https://github.com/user-attachments/assets/af9dcf72-ef82-43d0-ab57-27c64be6a161)|![Untitled](https://github.com/user-attachments/assets/0c5bfe75-3396-4c0a-979e-e949ee1194e0)|

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

Commit:
`Theme`: Use intersection observer to render a placeholder when not visible

Commit:
Update tests

Commit:
Update tests

Commit:
Add test

Commit:
E2E: wait for actionable theme to know the page is ready

We can't wait for the placeholders to be hidden anymore, since they are now used for lazy loading

Commit:
Fix test

Commit:
Increase the margin so a ful viewport is preloaded

Commit:
Fix test

It needed scrolling down until a premium theme was found

Commit:
Provide a `root` prop to the intersection observer

The logged in page wasn't aware of the incoming images because the document viewport is clipped by the theme showcase, so that needs to be the observer root element

Commit:
Don't pass the ref on the logged out theme showcase

BogdanUngureanu commented:
Thanks for the change.

Looks like there's still a regression on the Global View logged-in:
- Go to https://container-practical-cray.calypso.live/themes (Logged-in)
- Use a mobile resolution (e.g. iPhone 12)
- Refresh the page
- Scroll a bit
- Notice that the images are not loading.

Commit:
Use the document viewport as root on mobile devices

The showcase wrapper scrolls out of view when you down the page, so there's no intersection

xavier-lc commented:
@BogdanUngureanu nice catch! On mobile devices, the showcase wrapper scrolls out of view when you go down the page, so there's no intersection. I've fixed it so it uses the document viewport as root on mobile devices.

BogdanUngureanu commented:
Looks like there are still issues on mobile :(

https://github.com/user-attachments/assets/5f6d4bf5-33f1-417e-89d6-176ae5ffb367

Commit:
Add `isMobileWidthOrHeight` viewport util

Commit:
Update readme

Commit:
Check for mobile width or height

It's needed because in both cases the theme showcase wrapper ends up being out of view

Commit:
Update changelog

Commit:
Extract `ThemeShowcaseContextProvider`, the `ThemeShowcase` component doesn't really need to know the logic behind it

xavier-lc commented:
@BogdanUngureanu good catch! I've added a new [breakpoint util](https://github.com/Automattic/wp-calypso/pull/95984/commits/b1b9b09b1afca312cc49fdcf836bdc0549f9b942) because the layout rules apply both for mobile width or height.

BogdanUngureanu commented:
Thanks, @xavier-lc 

I did a new session and I've noticed two more issues :(
- We seem to make a new /plans request on each theme card display (on mobile)
- I'm seeing more theme thumbnail image flashes on mobile resolutions - every theme image is flashing on loading, giving a non-optimal scrolling UI.

https://github.com/user-attachments/assets/1530cee5-4e5b-4605-bd71-af9920b02d26

dsas commented:
Not blocking, just checking. It looks like this works on the logged-out theme showcase, and it works on the logged in global theme showcase (/themes) but it has no effect on logged-in site specific theme showcase (/themes/:siteSlug).

Happy for that to be a follow-up PR, but thought it was worth mentioning in case it's not what is expected.

Commit:
`usePlans`: Set `staleTime` to 1 hour

This prevents refetching it when lazy loading themes

xavier-lc commented:
> * We seem to make a new /plans request on each theme card display (on mobile) - this is more serious since it's happening on Logged out too (returning a lot of 403 error responses)

Addressed on https://github.com/Automattic/wp-calypso/pull/95984/commits/f9aebd88881124c108c67dfbbfa619343f35ae6e, preventing `usePlans` from being fired with every new set of themes.

Commit:
Add `MainRefContext`, a ref to the `Main` element is needed for lazy loading on mobile devices

Commit:
Adjust overflow and have it working on all scenarios

I hope 🙏

Commit:
Rename context hook

Commit:
Rename theme showcase context so it's clearer that it has to do with the intersection observer

mmtr commented on packages/data-stores/src/plans/queries/use-plans.ts:
```diff
@@ -73,6 +73,7 @@ function usePlans( {
 				} )
 			);
 		},
+		staleTime: 3600000, // 1 hour in ms
```
Why is this changed needed? Does this mean that if we ship some API changes to update the pricing or the name of some plans, they will not be applied for up to 1 hour? Because if that's the case, I think we need to discuss this change more deeply with other teams. With the current code, the expectation is that API changes to the plans will be reflected immediately in Calypso.

xavier-lc replied:
Bogdan mentioned that there were refetches when lazy loading a new batch of themes:

> We seem to make a new /plans request on each theme card display (on mobile) - this is more serious since it's happening on Logged out too (returning a lot of 403 error responses)

After some investigation I founds that it happens because the `ThemeTierUpgradeBadge` triggers a `usePlans` call when it's rendered, so that's what I set a `staleTime` of one hour.

> Does this mean that if we ship some API changes to update the pricing or the name of some plans, they will not be applied for up to 1 hour?

No, they would be applied immediately, but users that are already navigating the site without actually reloading it would not see the new data until one hour has gone by after the last hook call.

Anyway, the value can be set to one minute as well, I don't think a request per minute while the user scrolls the themes is something to be concerned. Wdyt?

mmtr replied:
> After some investigation I founds that it happens because the `ThemeTierUpgradeBadge` triggers a `usePlans` call when it's rendered, so that's what I set a `staleTime` of one hour.

This made me think 🤔 It seems that we're only interested in lazy loading the screenshots, right? So can we then move the lazy loading logic from the whole theme card to only the screenshots? That way the `ThemeTierUpgradeBadge` component (and the `usePlans` hook) won't be affected by these changes. 

I also wonder, why these changes increase the number of requests that `ThemeTierUpgradeBadge` component make to `/plans` from one per page load to one per each theme card? That might be a symptom of an underlying issue that we should fix (instead of masking it with a `staleTime`).

xavier-lc replied:
> So can we then move the lazy loading logic from the whole theme card to only the screenshots? That way the `ThemeTierUpgradeBadge` component (and the `usePlans` hook) won't be affected by these changes.

I thank that would still mean replacing the whole content rendered by `ThemeCard` with what's on the `renderPlaceholder` function of the `Theme` component. So the `badge` would be still re-rendered and triggering the `usePlans` hook.



> I also wonder, why these changes increase the number of requests that `ThemeTierUpgradeBadge` component make to `/plans` from one per page load to one per each theme card?

As I understand it, it's the expected `useQuery` behavior. Having the default `stleTime: 0` means that react query will always trigger a background request to the plans endpoint when a new instance of the `useQuery` hook is rendered, since the cached data is considered stale. That doesn't happen currently on production because all the themes are rendered at the same time and so at no point there's a new `usePlans` instance.

mmtr replied:
> I thank that would still mean replacing the whole content rendered by `ThemeCard` with what's on the `renderPlaceholder` function of the Theme component.

I meant the specific component used for the screenshot returned by the `renderScreenshot` function. But that made wonder: why are over-complicating this with an `InView` component when we already have a native solution with the `loading="lazy"` attribute that makes all the work? So I opened an alternative solution: https://github.com/Automattic/wp-calypso/pull/96961

mmtr reviewed (commented):
This tests well for me, but I'm bit concerned about the changes of the `usePlans` hook.

Also, as @dsas noted, theme screenshots are not lazy loaded in the site-specific TS (`/themes/:site`). I think that's expected and will be addressed in a follow-up, but has not been confirmed.

xavier-lc commented:
> This tests well for me, but I'm bit concerned about the changes of the `usePlans` hook.

Sure, I answered that on the comment.

> Also, as @dsas noted, theme screenshots are not lazy loaded in the site-specific TS (`/themes/:site`). I think that's expected and will be addressed in a follow-up, but has not been confirmed.

Yeah, I didn't look into that because the PR had enough changes without it, I'll tackle it on a followup PR.

mmtr commented:
Closing this in favor of https://github.com/Automattic/wp-calypso/pull/96961
